Add Pomelo.EntityFrameworkCore.MySql component

bgrainger opened the pull request:
Follow up to #825 that adds the related EF Core component for MySQL. Fixes #786.

This will need the following NuGet packages to be mirrored:

* MySqlConnector.Logging.Microsoft.Extensions.Logging 2.1.0
* Pomelo.EntityFrameworkCore.MySql 8.0.0-beta.2

Possible areas of concern:

* Pomelo needs a `ServerVersion` passed in; this PR always uses a `string` so it can be bound from configuration
* Pomelo can't use `MySqlDataSource` so this configures the DbContext with a connection string
* Because `MySqlDataSource` isn't used, we have to configure logging through the "legacy" MySqlConnector API
* These last two may need fixes in Pomelo: https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1817. CC @lauxjpn

bgrainger committed:
Add Pomelo.EntityFrameworkCore.MySql component.

bgrainger committed:
Add Pomelo.EntityFrameworkCore.MySql tests.

lauxjpn commented:
We are adding support for `DbDataSource` with https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/pull/1819 (will be merged tomorrow).

(The next release will be `8.0.0-silver.1`. It should be on NuGet around the 2023-12-09 weekend.)

eerhardt reviewed (commented):
Thanks for another contribution, @bgrainger

I have added the 2 new nuget packages to the internal mirror. They should be finished mirroring in a few minutes.

bgrainger commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -0,0 +1,163 @@
+                    // add metrics from the underlying MySqlConnector ADO.NET library
+                    meterProviderBuilder.AddMeter("MySqlConnector");
+
+                    // TODO: add any Pomelo EF Core metrics
```
There isn't one. Pomelo doesn't add its own metrics on top of EF Core and I don't think they have any plans to add any. So this was a pointless comment and I'll delete it.

bgrainger committed:
Merge main into pomelo.

Conflicts:
	Aspire.sln

bgrainger commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -0,0 +1,163 @@
+                MySqlConnectorLogManager.Provider = new MicrosoftExtensionsLoggingLoggerProvider(loggerFactory);
+            }
+
+            if (settings.ServerVersion is null)
+            {
+                throw new InvalidOperationException($"ServerVersion is missing. It should be provided in the ServerVersion' key in '{DefaultConfigSectionName}' or '{typeSpecificSectionName}' configuration section, or set by the 'configureSettings' callback.");
+            }
```
@lauxjpn Is there any way for a default value to work here, or is it absolutely required to be set by the user?

bgrainger replied:
Using `ServerVersion.AutoDetect` will immediately open a database connection to query the server's version and parse it. This may be undesirable in the `ConfigureDbContext` code?

However, it's probably a better default than throwing an exception; users can opt into specifying a server version to optimize this process or avoid unexpected startup failures (e.g., if the DB isn't available).

bgrainger replied:
See https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/blob/5fa77d7f68cc51ea6e5d1206fd2cbf58d20f3182/src/EFCore.MySql/Infrastructure/ServerVersion.cs#L71-L77.

bgrainger replied:
I don't know enough about Pomelo to answer this, but I am now curious if `AutoDetect` could be a "sentinel" value that simply prompts Pomelo to check and set its internal ServerVersion property when a connection is first opened to a database. (That would need to happen before the first SQL statement that depends on the target server version is created, but doesn't seem out of the realm of possibility.) CC @lauxjpn

eerhardt replied:
Why doesn't MySqlConnector need this? Why is it able to establish a connection to the MySql server without having to specify the version?

bgrainger replied:
MySqlConnector is a low-level ADO.NET library that just speaks the MySQL protocol. It doesn't really care (that much) about the contents of the packets it transmits over the wire, or what the destination server is (as long as it's compatible with the MySQL protocol).

Because Pomelo creates and executes DDL statements (to implement EF Core), it's much more sensitive to which server is on the other end. For example, MySQL supports the `JSON` column type, but MariaDB doesn't and treats it as an alias for `LONGTEXT`. Or CTEs are only supported in MySQL >= 8.x and MariaDB >= 10.2 so (I assume) Pomelo uses CTEs when they're available, but has to emulate them with different SQL syntax on servers that don't support them.

bgrainger replied:
It looks like Pomelo sets up some data type mappings before the first connection is established (e.g., setting whether `DATETIME(6)` can be used or whether it has to fall back to `DATETIME` for servers that don't support precision). That probably makes my "`AutoDetect` is a sentinel" idea harder to implement, but maybe still possible?

It looks like there's some explanation of the current behaviour here: https://github.com/PomeloFoundation/Pomelo.EntityFrameworkCore.MySql/issues/1603#issuecomment-1016922466. I'm not sure I completely agree with the rationale; it feels like Pomelo would be easier to use for most people if it "just worked" by default (via auto-detection) and people who care about determinism or low overhead could opt into that.

lauxjpn replied:
If you want to use `AutoDetect`, then you should probably use it together with Polly. Otherwise, app startup might just fail if the network hiccups.

---

We usually discourage the use of the `AutoDetect` helper method in production code, because it makes users oblivious to differences in the implementation of MySQL compatible database servers and versions. Depending on the configured database version, Pomelo uses different syntax options and database features. If you have a database version on your development machine that differs from the one on your production system, your locally working code might throw exceptions at runtime due to unavailable features or an unsupported SQL syntax.

It is therefore best practice to explicitly specify the server version or at least to explicitly opt-in to using auto-detect (to ensure that someone doing that is aware of the implications).

Using `AutoDetect` as a default is nice for demos, but dangerous for developing actual production systems. If Aspire is targeted at real-world production systems, `AutoDetect` should not be the default.

bgrainger replied:
> Depending on the configured database version, Pomelo uses different syntax options and database features. If you have a database version on your development machine that differs from the one on your production system, your locally working code might throw exceptions at runtime due to unavailable features or an unsupported SQL syntax.

Wouldn't this problem be exacerbated by hard-coding the `ServerVersion`? 

That is, if I hardcoded `new MySqlServerVersion("8.0.21")` in my code, Pomelo will generate SQL that uses the `JSON_VALUE` function. If my production server is 8.0.20, this would fail when deployed to production.

But if I used `ServerVersion.AutoDetect`, Pomelo would generate the appropriate SQL syntax based on the server it's communicating with. 

That is, I'm having a hard time not understanding `AutoDetect` as being more resilient (albeit less predictable). I may have to think about this some more.

bgrainger committed:
Remove incorrect documentation.

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

bgrainger committed:
Remove irrelevant TODO comments.

bgrainger committed:
Update XML documentation.

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

bgrainger committed:
Update XML documentation to current conventions.

bgrainger committed:
Add documentation for Pomelo metrics.

bgrainger commented on tests/Aspire.Pomelo.EntityFrameworkCore.MySql.Tests/AspireEFMySqlExtensionsTests.cs:
```diff
@@ -0,0 +1,120 @@
+// Licensed to the .NET Foundation under one or more agreements.
```
I'm not sure I understand the ask here. #1217 seems to add tests for Aspire.Hosting. That project (and Aspire.Hosting.Tests) has no dependency (and should have no dependency) on Pomelo.EFCore.MySql (from this PR).

Are there more MySQL functional tests you would like to see added to Aspire.Hosting.Tests (that are logically more of a follow-up to #825)?

bgrainger replied:
Sorry for being dense, but I'm still not understanding what's needed here.

How is the MySQL scenario (MySqlConnector vs Pomelo.EntityFrameworkCore.MySql) different from PostgreSQL (Npgsql vs Npgsql.EntityFrameworkCore.PostgreSQL)? I don't see any EF-specific tests in Aspire.Hosting.Tests so it's still not clear to me what a test for the second runtime component would look like. Can you point to a Npgsql or SQL Server example that can be used as a template?

(Just as Npgsql.EntityFrameworkCore.PostgreSQL depends on Npgsql for its DB connectivity, Pomelo.EntityFrameworkCore.MySql depends on MySqlConnector, so if MySqlConnector can successfully connect to a test Docker container, Pomelo EFCore should be able to, too.)

This PR adds the Aspire.Pomelo.EntityFrameworkCore.MySql.Tests project, which I thought was sufficient for testing Aspire.Pomelo.EntityFrameworkCore.MySql.

eerhardt replied:
> How is the MySQL scenario (MySqlConnector vs Pomelo.EntityFrameworkCore.MySql) different from PostgreSQL (Npgsql vs Npgsql.EntityFrameworkCore.PostgreSQL)? 

It isn't different.

> I don't see any EF-specific tests in Aspire.Hosting.Tests

These tests are very new, and not yet complete. I've opened https://github.com/dotnet/aspire/issues/1566 for them.

> so it's still not clear to me what a test for the second runtime component would look like.

It would look the same as the non-EF version. Only instead of using the non-EF component, it would use the EF version. The "hosting" side of the test would still be the exact same - we only need 1 MySQL server/container running. But we want to verify that each components work end-to-end with the Hosting pieces. We've had some early bugs that would have been caught if we had these tests to start with.

> This PR adds the Aspire.Pomelo.EntityFrameworkCore.MySql.Tests project, which I thought was sufficient for testing Aspire.Pomelo.EntityFrameworkCore.MySql.

They are a good start, but it doesn't verify that the Aspire.Pomelo.EntityFrameworkCore.MySql component works with the MySql hosting pieces.

bgrainger replied:
Thanks for the pointers; I think I finally understand what you're suggesting and have added dedicated integration tests for the Pomelo component in 817940ba30f074b84cc8d6204055c635b2581c62.

bgrainger committed:
Change catch block to the exception Pomelo throws.

eerhardt reviewed (commented):
Just a few comments. Once they are addressed I think this is ready to be merged.

bgrainger committed:
Merge main into pomelo.

Conflicts:
	Aspire.sln
	src/Components/Aspire_Components_Progress.md
	src/Components/Telemetry.md

bgrainger committed:
Reuse existing TestDbContext.cs file.

bgrainger committed:
Use configuration schema generator.

bgrainger committed:
Remove "Container" from AddMySql.

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

bgrainger committed:
Allow ServerVersion setting to be optional.

bgrainger committed:
Remove (now optional) ServerVersion setting from test.

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AssemblyInfo.cs:
```diff
@@ -0,0 +1,7 @@
+using Aspire;
+using Aspire.Pomelo.EntityFrameworkCore.MySql;
+
+[assembly: ConfigurationSchema("Aspire:Pomelo:EntityFrameworkCore:MySql", typeof(PomeloEntityFrameworkCoreMySqlSettings))]
```
We should also add all the logging categories that are in Telemetry.md. That way they light up in the appsettings.json file intellisense.

bgrainger replied:
I think I shouldn't have duplicated the MySqlConnector logging categories into the Pomelo section in Telemetry.md. (I assume they will ultimately get pulled in through being a transitive dependency?) I'll remove them from Telemetry.md.

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -0,0 +1,165 @@
+                MySqlConnectorLogManager.Provider = new MicrosoftExtensionsLoggingLoggerProvider(loggerFactory);
+            }
+
+            // We don't register logger factory, because there is no need to: https://learn.microsoft.com/dotnet/api/microsoft.entityframeworkcore.dbcontextoptionsbuilder.useloggerfactory?view=efcore-7.0#remarks
```
This comment is kind of weird given that directly above it we are explicitly registering a logger factory.

bgrainger replied:
Was copy/pasted; will delete.

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/Aspire.Pomelo.EntityFrameworkCore.MySql.csproj:
```diff
@@ -0,0 +1,24 @@
+  </ItemGroup>
+
+  <ItemGroup>
+    <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks.EntityFrameworkCore" />
```
```suggestion
    <PackageReference Include="Microsoft.Extensions.Configuration.Binder" />
    <PackageReference Include="Microsoft.Extensions.Diagnostics.HealthChecks.EntityFrameworkCore" />
```

We need to explicitly reference the `Microsoft.Extensions.Configuration.Binder` package in order to "up lift" the version to 8.0, which has the source generator in it. I am thinking of how we can do this generally for all components, so the next one doesn't need this explicit line.

bgrainger commented:
I think I have time to look at this more next week. Is there a hard cut-off date to make it into Preview 3?

Did we land in a "good enough" place with `ServerVersion.AutoDetect`? I'm ultimately not really happy with how it's implemented (in Pomelo) but that seems out-of-scope for this PR (to me).

Then the remaining big component is "tests that test the end-to-end of both the hosting and the runtime components" (plus a few other minor still-unresolved comments you left)?

Thanks for all the help and feedback with this PR so far!

davidfowl commented:
Are we getting this for preview3?

bgrainger committed:
Merge main into pomelo.

Conflicts:
	Aspire.sln
	src/Components/Aspire_Components_Progress.md

bgrainger committed:
Fix bad merge.

bgrainger committed:
Remove MySqlConnector logging categories from Pomelo section.

bgrainger committed:
Delete meaningless comment.

bgrainger committed:
Add Microsoft.Extensions.Configuration.Binder.

Co-authored-by: Eric Erhardt <eric.erhardt@microsoft.com>

bgrainger committed:
Revert "Remove (now optional) ServerVersion setting from test."

This reverts commit f1334520641be62d3af5a52b7f77919c075641d3.

bgrainger committed:
Add basic MySQL Aspire.Hosting tests.

These were modeled on AddPostgresTests and AddRedisTests.

bgrainger committed:
Add Pomelo EFCore integration tests.

bgrainger committed:
Merge main into pomelo.

bgrainger committed:
Incorporate project change from dff1467f.

bgrainger commented:
I've added what I think the Pomelo EFCore E2E tests should look like. Since these are the first EF E2E tests (AFAICT) I may have approached it differently from what you expected. Let me know if you want any changes there.

This PR is ready for review.

eerhardt commented on src/Components/Aspire.Pomelo.EntityFrameworkCore.MySql/AspireEFMySqlExtensions.cs:
```diff
@@ -0,0 +1,164 @@
+                {
+                    ThrowForMissingConnectionString();
+                }
+                serverVersion = ServerVersion.AutoDetect(connectionString);
```
Playing with this change locally, I'm a bit concerned about this now. I'm trying to run the integration tests (which aren't run in CI yet - we are working on that). If the Pomelo test is run by itself, it fails because just trying to get the DbContext out of DI is trying to establish a connection. But the MySql docker container isn't up yet (it takes a while to get up) and this line is throwing because it can't make a connection.

I wonder if we should retry this up to the MaxRetryCount. It isn't a great experience to fail getting the DbContext because it happens as part of ASP.NET calling into the user's code. So there's no chance for them to retry.

Thoughts? Maybe this shouldn't block the initial check-in, since you can work around the issue by setting ServerVersion explicitly.

bgrainger replied:
> I wonder if we should retry this up to the MaxRetryCount. 

And then still fail with a fatal error if it can't make a connection? (As opposed to making up a value like `"8.0.0"` or `"5.7.28"` or something that might not actually be safe if the user is running MariaDB, depending on how much this changes Pomelo's behavior.)

That definitely seems like an improvement over the status quo to me. (Unless of course 100% of the time it still fails but now just takes longer.)

In my original PR (before 2623f9322a6f81a22eb90ba0765f9f274f18258b), this was a mandatory option that had to be supplied by the user. That was less user-friendly on one hand, but maybe more predictable?

Very random thought (and may be going completely down the wrong path): is there some "magic" we could do by detecting if there's a `ContainerImageAnnotation`? E.g., set an appropriate default ServerVersion by faking a version string based on whether the user is using `mysql:latest` or `mariadb:11.0`? (That's not actually customisable in `AddMySqlContainer` right now though maybe it should be.) That only works for the local Docker scenario, but maybe that's OK because for production deployment we can assume that the "real" database is up and serving requests?

eerhardt replied:
Let's discuss the options in a new issue explicitly for this. Can you open it? or do you want me to?

bgrainger replied:
https://github.com/dotnet/aspire/issues/1866

eerhardt committed:
PR feedback

- Add log categories to ConfigurationSchema
- Check the "API" boxes for the MySql components in the Progress doc
- Issue a query in the functional test. Follow the same pattern as Oracle
- Set the ServerVersion explicitly in the functional test, so the connection can be retried when the MySql server isn't running.

eerhardt commented:
This is looking great, @bgrainger. Thanks for another great contribution.

I had a few tweaks that I made in https://github.com/mysql-net/aspire/pull/1 (I couldn't push to your branch directly). Once those changes are made, I think this will be ready to merge. We can log a follow up issue on my [comment above](https://github.com/dotnet/aspire/pull/1161#discussion_r1466887228).

bgrainger commented:
> Since these are the first EF E2E tests (AFAICT)

Well, not sure how I missed the Oracle ones, but now I see them plain as day! 

Thanks for pushing the fixes directly; pulled that into this branch.

eerhardt reviewed (approved):
This looks great. I'll merge this either tonight or tomorrow to give other people time to review.
